manager/controllers/step_controller.go

```diff
@@ -131,6 +131,44 @@ func (r *StepReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.
 
 	ownerReferences := []metav1.OwnerReference{*metav1.NewControllerRef(step.GetObjectMeta(), dfv1.StepGroupVersionKind)}
 
+	serviceName := pipelineName + "-" + stepName
```

Comment:
my suggestion could be this, we always create a service named `pipelineName- stepName`, and, if the source has another service name, we created that as well

Reply:
Sounds good.

Reply:
Just realized headless service can not do port forward, it always use the `targetPort`. In this case, there are 2 options:

1. Change 3570 to 443;
2. Create a headless service and a regular service.

Which one do you like, or do you have any other idea? @alexec

Reply:
Why do you want to do headless? Is normal OK?

Reply:
Normal service does not work.

Reply:
the only purpose of the headless service is to find out how many replicas?

Reply:
Does it seems like misuse? Is there another way, e.g. Kubernetes API request?

Reply:
The purpose of headless service:

1. Used by the pod to find out how many replicas;
2. Controller can easily find the PODs' metrics endpoints without querying pod IPs through k8s api.

Reply:
By "controller" you mean Prometheus. I think (controller has an informer and know how many pods there are). 

I'm starting to wonder if the controller should return this metric instead?

Reply:
No. Since we are gonna stop reporting information like `pending` from the pod the step status, the controller will not check step status for scaling, instead, it will access pod metrics api directly for that. So the controller need to have a easily way to get the pod metrics endpoints, using a DNS will be a good way without querying pod IPs through k8s api.